File: dot_prod/dot_prod_mpi_noreduce.f90
! -*- Mode: Fortran90; -*-
!-----------------------------------------------------------------
! Daniel R. Reynolds
! SMU, Mathematics
! Math 4370/6370
! 7 February 2015
!=================================================================


program DotProd_MPI
  !-----------------------------------------------------------------
  ! Description: 
  !    Computes the dot-product of two vectors.
  !-----------------------------------------------------------------
  !======= Inclusions ===========
  use mpi
!  include "mpif.h"  ! include this if mpi module unavailable

  !======= Declarations =========
  implicit none

  integer, parameter :: MAX_INTEGER_DIGITS = 10
  integer :: n, p, i, is, ie, numprocs, myid, ierr
  integer :: stat(MPI_STATUS_SIZE)
  character(MAX_INTEGER_DIGITS) :: n_string
  double precision, allocatable :: a(:), b(:)
  double precision :: mysum, sum, stime, ftime
  double precision :: alloctime, inittime, runtime

  !======= Internals ============
  
  ! intialize MPI
  call MPI_Init(ierr)
  if (ierr /= 0) then
     write(0,*) ' error in MPI_Init =',ierr
     stop
  endif
  call MPI_Comm_size(MPI_COMM_WORLD, numprocs, ierr)
  if (ierr /= 0) then
     write(0,*) ' error in MPI_Comm_size =',ierr
     call MPI_Abort(MPI_COMM_WORLD, 1, ierr)
  endif
  call MPI_Comm_rank(MPI_COMM_WORLD, myid, ierr)
  if (ierr /= 0) then
     write(0,*) ' error in MPI_Comm_rank =',ierr
     call MPI_Abort(MPI_COMM_WORLD, 1, ierr)
  endif

  ! get n from the command line
  call getarg(1, n_string)

  ! ensure that an argument was passed in
  if ( n_string == '' ) then
     write(0,*) 'Error: function requires one argument (vector length)'
     call MPI_Abort(MPI_COMM_WORLD, 1, ierr)
  endif
  
  ! convert n_string to integer, and ensure it's positive
  read (n_string, *) n
  if ( n < 1 ) then
     write(0,*) 'Error: vector length must be greater than 0'
     call MPI_Abort(MPI_COMM_WORLD, 1, ierr)
  endif


  ! determine this processor's interval
  is = floor(1.d0*n/numprocs)*myid + 1
  ie = floor(1.d0*n/numprocs)*(myid+1)
  if (myid == numprocs-1)  ie = n

  ! allocate the local portion of vectors
  stime = MPI_Wtime()
  allocate(a(is:ie),b(is:ie))
  ftime = MPI_Wtime()
  alloctime = ftime - stime

  ! initialize the vector values
  stime = MPI_Wtime()
  do i=is,ie
     a(i) = 1.d-3*i/n
     b(i) = 1.d-3*(n-i)/n
  enddo
  ftime = MPI_Wtime()
  inittime = ftime - stime

  ! start computation timer
  stime = MPI_Wtime()

  ! perform dot-product
  mysum = 0.d0
  do i=is,ie,1
     mysum = mysum + a(i)*b(i)
  enddo

  ! perform manual reduction
  if (myid /= 0) then
     ! everyone sends value to root proc
     call MPI_Send(mysum, 1, MPI_REAL8, 0, myid, MPI_COMM_WORLD, ierr)
     if (ierr /= MPI_SUCCESS) then
        write(0,*) 'Error in MPI_Send'
        call MPI_Abort(MPI_COMM_WORLD, 1, ierr)
     endif
  else
     ! root receives values from others and adds to own
     sum = mysum
     do p=1,numprocs-1
        call MPI_Recv(mysum, 1, MPI_REAL8, MPI_ANY_SOURCE, &
                      MPI_ANY_TAG, MPI_COMM_WORLD, stat, ierr)
        if (ierr /= MPI_SUCCESS) then
           write(0,*) 'Error in MPI_Recv'
           call MPI_Abort(MPI_COMM_WORLD, 1, ierr)
        endif
        sum = sum + mysum
     enddo
  endif

  ! stop timer
  ftime = MPI_Wtime()
  runtime = ftime-stime

  ! output computed value and runtime
  if (myid == 0) then
     print *, ' vector length =',n
     print *, '   dot-product =',sum
     print *, '    alloc time =',alloctime
     print *, '     init time =',inittime
     print *, '      run time =',runtime
  endif

  ! free vectors
  deallocate(a,b)

  ! finalize MPI
  call MPI_Finalize(ierr)  

end program DotProd_MPI
!=================================================================

File: dot_prod/dot_prod_mpi_noreduce.c
/* Daniel R. Reynolds
   SMU Mathematics
   Math 4370/6370
   7 February 2015 */

/* Inclusions */
#include <stdlib.h>
#include <stdio.h>
#include "mpi.h"


/* Example routine to compute the dot-product of two vectors */
int main(int argc, char* argv[]) {

  /* declarations */
  int i, n, p, is, ie;
  double *a, *b, sum, mysum, alloctime, inittime, runtime;
  double stime, ftime;
  int numprocs, myid;
  MPI_Status stat;

  /* intialize MPI */
  if (MPI_Init(&argc, &argv) != MPI_SUCCESS) {
    fprintf(stderr, "Error in MPI_Init\n");
    return 1;
  }
  if (MPI_Comm_size(MPI_COMM_WORLD, &numprocs) != MPI_SUCCESS) {
    fprintf(stderr, "Error in MPI_Comm_size\n");
    MPI_Abort(MPI_COMM_WORLD, 1);
  }
  if (MPI_Comm_rank(MPI_COMM_WORLD, &myid) != MPI_SUCCESS) {
    fprintf(stderr, "Error in MPI_Comm_rank\n");
    MPI_Abort(MPI_COMM_WORLD, 1);
  }

  /* ensure that an argument was passed in */
  if (argc < 2) {
    fprintf(stderr,"Error: function requires one argument (vector length)\n");
    MPI_Abort(MPI_COMM_WORLD, 1);
  }

  /* set n as the input argument, and ensure it's positive */
  n = atoi(argv[1]);
  if (n < 1) {
    fprintf(stderr,"Error: vector length %i must be greater than 0\n", n);
    MPI_Abort(MPI_COMM_WORLD, 1);
  }

  /* determine this processor's interval */
  is = ((int) (1.0*n/numprocs))*myid;
  ie = ((int) (1.0*n/numprocs))*(myid+1);
  if (myid == numprocs-1)  ie = n;

  /* allocate the local portion of vectors */
  stime = MPI_Wtime();
  a = (double *) malloc( (ie-is)*sizeof(double) );
  b = (double *) malloc( (ie-is)*sizeof(double) );
  ftime = MPI_Wtime();
  alloctime = ftime - stime;

  /* initialize the vector values */
  stime = MPI_Wtime();
  for (i=is; i<ie; i++) {
    a[i-is] = (0.001 * (i + 1.0)) / n;
    b[i-is] = (0.001 * (n - i - 1.0)) / n;
  }
  ftime = MPI_Wtime();
  inittime = ftime - stime;

  /* start computation timer */
  stime = MPI_Wtime();

  /* compute dot-product */
  mysum = 0.0;
  for (i=0; i<ie-is; i++)  mysum += a[i]*b[i];

  /* perform manual reduction */
  if (myid != 0) {
    /* everyone sends value to root proc */
    if (MPI_Send(&mysum, 1, MPI_DOUBLE, 0, 
		 myid, MPI_COMM_WORLD) != MPI_SUCCESS) {
      fprintf(stderr,"Error in MPI_Send\n");
      MPI_Abort(MPI_COMM_WORLD, 1);
    }
  } else {
    /* root receives values from others and adds to own */
    sum = mysum;
    for (p=1; p<numprocs; p++) {
      if (MPI_Recv(&mysum, 1, MPI_DOUBLE, MPI_ANY_SOURCE, MPI_ANY_TAG, 
		   MPI_COMM_WORLD, &stat) != MPI_SUCCESS) {
	fprintf(stderr,"Error in MPI_Send\n");
	MPI_Abort(MPI_COMM_WORLD, 1);
      }
      sum += mysum;
    }
  }

  /* stop timer */
  ftime = MPI_Wtime();
  runtime = ftime-stime;

  /* output computed value and runtime */
  if (myid == 0) {
    printf(" vector length = %i\n",n);
    printf("   dot-product = %.16e\n",sum);
    printf("    alloc time = %.2e\n",alloctime);
    printf("     init time = %.2e\n",inittime);
    printf("      run time = %.16e\n",runtime);
  }

  /* free vectors */
  free(a);
  free(b);

  /* finalize MPI */
  MPI_Finalize();

} /* end main */

